src/NHibernate/Linq/Visitors/ExpressionKeyVisitor.cs

```diff
@@ -109,19 +109,23 @@ protected override Expression VisitConstant(ConstantExpression expression)
 				{
 					_string.Append("NULL");
 				}
-				else
+				else if (expression.Value is IQueryable)
 				{
-					var value = expression.Value as IEnumerable;
-					if (value != null  && !(value is string) && !(value is IQueryable))
-					{
-						_string.Append("{");
-						_string.Append(String.Join(",", value.Cast<object>()));
-						_string.Append("}");
-					}
-					else
+					_string.Append(expression.Value);
+				}
+				else if (expression.Value is IEnumerable enumerable && !(enumerable is string))
+				{
+					_string.Append("{");
+					foreach (var value in enumerable)
 					{
-						_string.Append(expression.Value);
+						_string.Append(value).Append("#").Append(value.GetHashCode()).Append(",");
```

Comment:
I have a hard time conceiving a valid query having objects instances instead of primitive types in a list/array. Are you sure this is useful to change the enumerable case too?